### PR TITLE
util/watch: Move event handling into its own function

### PR DESCRIPTION
**Pull request by sharnoff**
This also fixes a possible race on update events, because we weren't locking the store.

The function basically already existed by itself in the body of Watch as an anonymous function. Other than moving the locks, no other changes were made.

I think something very similar was originally implemented in #65, but that PR seems mostly stalled.

This PR is mostly "cleaning things up a little as I'm figuring out where to add metrics".

### CODE REVIEW

File: pkg/util/watch.go [hunk 1]
```diff
@@ -493,6 +428,82 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
...
+	eventType watch.EventType,
+	meta metav1.Object,
+	ptr P,
+) error {
+	uid := meta.GetUID()
+	name := NamespacedName{Namespace: meta.GetNamespace(), Name: meta.GetName()}
+	obj := (*T)(ptr)
+
+	// Some of the cases below don't actually require locking the store. Most of the events that we
+	// recieve *do* though, so we're better off doing it here for simplicity.
```
Comment (tychoish) [inline pkg/util/watch.go:445]: could have each case in the switch statement be: 
```go
return func() error {
// lock handling
// current code
}
```
I think it'd be worth it, given how contended this lock may be.

File: pkg/util/watch.go [hunk 2]
```diff
@@ -493,6 +428,82 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 	return &store, nil
 }
 
+// helper for Watch. Error events are expected to already have been handled by the caller.
+func handleEvent[T any, P ~*T](
```
Comment (tychoish) [inline pkg/util/watch.go:432]: wasn't this my handcrank change, more or less?

**Review by tychoish — APPROVED** [review]
love it. 

I'm worried enough about lock contention in this codebase that I'd move the lock around.

**Review by tychoish — APPROVED** [review]
love it. 

I'm worried enough about lock contention in this codebase that I'd move the lock around.

File: pkg/util/watch.go (hunk 2 above)
Comment (sharnoff) [reply pkg/util/watch.go:432]: yep!

> I think something very similar was originally implemented in https://github.com/neondatabase/autoscaling/pull/65, but that PR seems mostly stalled.

File: pkg/util/watch.go (hunk 1 above)
Comment (sharnoff) [reply pkg/util/watch.go:445]: Realistically we'll be fine. The only "normal" case that won't require locking is if we receive a "bookmark" event, which shouldn't be very often. (the other cases are things that we're not expecting, so will panic)

**Pull request merged**